### ISSUE TITLE
Item pipes don't work properly

### ISSUE THREAD

**Issue by AlexIIL** [AlexIIL/Buildcraft_1.8_Issues#55]
- Items go further than they should
   (They go out WAAY further and then reset back)
- Pipes can protect against explosions, but die as a result

**Comment by LuigiHutch** [comment]
Also the Obsidian pipes does not work properly:
![2016-01-11_22 53 37](https://cloud.githubusercontent.com/assets/9326429/12249287/af9d75fa-b8b6-11e5-9a24-91ef4a798d4a.png)

**Comment by MoreThanHidden** [comment]
> Also the Obsidian pipes does not work properly:

This is a separate issue but looks like you haven't got a lever on that engine, the obsidian pipes seem to be working for me.

**Comment by LuigiHutch** [comment]
> This is a separate issue but looks like you haven't got a lever on that engine, the obsidian pipes seem to be working for me.

The engine is recieving a redstone signal via the gate. They still dont work properly, Steps to reproduce:
1. Place down an obsidian Pipe
2. Throw items in front of it
3. Now power it, I powered it with an engine for an example.

The items wont get picked up. This was tested again in BC pre-10

**Comment by MoreThanHidden** [comment]
Not sure if something has changed in recent commits but in dev:
![obsidianpipe](https://cloud.githubusercontent.com/assets/1351964/12872053/26d09e48-cdd1-11e5-9afa-4969d4db34fa.gif)

**Comment by LuigiHutch** [comment]
Maybe, well all I know it it isn't occurring in BC pre-10

**Comment by AlexIIL** [comment]
Obsidian pipes were fixed after pre10

**Comment by MrGraphic** [comment]
Just my two cents on first issue. I belive that @MoreThanHidden's solution is insufficient. With that big number ( 0.3 ) items will never reach end of the pipe and simply "jumps" during transition. Also it can not be a constant as items in pipes can move with diffrent speeds ( that's why it was multiplied in original code probably ). 
I've made small change to that part ( forgive me, last time I wrote any code was eons ago ) and it seems to be smoother but still not perfect ( still choppy at slowest speed in horizontal movement )

`return Utils.convertMiddle(container.getPos()).subtract(item.pos).lengthVector() <= (0.24900001 + item.getSpeed()/10);`

was:

`return Utils.convertMiddle(container.getPos()).subtract(item.pos).lengthVector() < 0.3;`

I know you guys know better how to do this in the correct way but I thought I will add something to improve aesthetics while there are more important fixes on @AlexIIL's head :).

**Comment by MoreThanHidden** [comment]
In my tests this was working fine as this is a check point, to work out if the item has past the centre when it hits the end of a pipe. Changing this value shouldn't do anything aesthetically? 

I'll have another look after work.

**Comment by MrGraphic** [comment]
It is safe to say that I have small idea what I'm doing in this case hehe ;). I just noticed that with 0.3 items' movment is very choppy. Item reaches like 95% of the way then jumps to another pipe/block. The lower the value ( minimum just above 0.25 ), the smother the tranzit ( only with item speed higher than 0.1 - lowest value ).

**Comment by MoreThanHidden** [comment]
OK well I have finally achieved a perfect transition:

```
    protected boolean middleReached(TravelingItem item) {  
    float middleLimit = item.getSpeed() * 1.01F;  
    Vec3 middlepoint = new Vec3(0.5, TransportUtils.getPipeFloorOf(item.getItemStack()),0.5);  
    Vec3 compare = Utils.convert(container.getPos()).add(middlepoint).subtract(item.pos);  
        return Math.abs(compare.xCoord) &lt; middleLimit  
                && Math.abs(compare.yCoord) &lt; middleLimit  
                && Math.abs(compare.zCoord) &lt; middleLimit;  
    }
```

End Reached also needs to be changed to make it smooth:

```
protected boolean endReached(TravelingItem item) {
        return item.pos.distanceTo(Utils.convertMiddle(container.getPos())) > 0.65;
    }
```

**Comment by MrGraphic** [comment]
Yeah, no idea what I'm doing :D but... I tweaked one more value based on what I saw looking at item position values during movment:

`protected boolean endReached(TravelingItem item) {
        return item.pos.distanceTo(Utils.convertMiddle(container.getPos())) > 0.7;`
Was 0.5 before and now item movment is soo smooth :D. 
I have no idea if this will cause ( or not ) Stev's head explosion but it works - except for motion with vector 0,-1,0. No idea how to fix this one direction :|.

**Comment by MrGraphic** [comment]
Hahaha look at that ;)

**Comment by MoreThanHidden** [comment]
Beat you by 3 seconds :stuck_out_tongue_winking_eye:

**Comment by MrGraphic** [comment]
:stuck_out_tongue: 
Was going to check but unfortunelly I'm unable to compile that code of yours.

**Comment by MoreThanHidden** [comment]
Oh the comment  has messed with it I'll sync the PR shortly.

**Comment by MrGraphic** [comment]
Quick question: does that code handles also that downwards movment "jumping" ?

Now I noticed what git did with "<" and ">" symbols hehe

**Comment by MoreThanHidden** [comment]
> Quick question: does that code handles also that downwards movment "jumping" ?

Yes - `TransportUtils.getPipeFloorOf(item.getItemStack())`

**Comment by MrGraphic** [comment]
Almost perfect @MoreThanHidden :) but now high speed items with vector 1,0,0 are having minor problems - they are dissapearing visually.
Now I can watch at networks with gazilions items without getting hiccups :).

**Comment by MrGraphic** [comment]
@MoreThanHidden @AlexIIL I saw conversation under under PR and I think there is easy solution for that by switching to dynamic values:

``` java
protected boolean endReached(TravelingItem item) {
return item.pos.distanceTo(Utils.convertMiddle(container.getPos())) > 0.71 - item.getSpeed()/TransportConstants.PIPE_MAX_SPEED*0.2;
```

While this is not a fix for the core of the problem it helps with like 99% of the cases of vanishing items.
Why "0.7" and not "0.65" ? From my observations it is simply smoother for the minimum speed of items.
